File: dev/front/src/types/mcp.ts
/**
 * MCP 资源类型镜像(唯一事实源:API-CONTRACT.md §2.6)。
 *
 * 后端 `model_dump(exclude_none=True)`:未设置的字段不会出现(整个键缺失),
 * 故可选字段全部以 `?:` 声明,渲染时按可选处理。
 */

/** stdio 传输:command 必填,args/env 可选。 */
export interface StdioServer {
  transport: 'stdio'
  command: string
  args?: string[]
  env?: Record<string, string>
}

/** http 传输:url 必填,headers 可选。 */
export interface HttpServer {
  transport: 'http'
  url: string
  headers?: Record<string, string>
}

/** MCP 服务器配置判别联合:类型层按 transport 收窄。 */
export type MCPServerConfig = StdioServer | HttpServer

/** GET /mcp/servers 响应:name → config 字典。 */
export interface MCPServerListResponse {
  servers: Record<string, MCPServerConfig>
}

/** POST /mcp/servers 与 DELETE /mcp/servers/{name} 响应。 */
export interface MCPServerOpResult {
  ok: boolean
  name: string
}

/**
 * POST /mcp/servers/{name}/test 响应。
 * `ok: false` 是正常业务返回(探测降级,不抛),UI 显示"未连通"而非红色报错。
 */
export interface MCPTestResult {
  ok: boolean
  name: string
  /**
   * B8 落地前后端未归一化:可能是 string[],也可能是对象数组;
   * 渲染层需防御式处理(见契约 §2.6 的 label 取值示例)。
   */
  tools: unknown[]
}

File: src/api/routers/mcp.py
"""MCP 服务器管理路由(模块 11 T3):配置 CRUD + 连通性测试。

复用 08 的 config(load/add/remove)与 client(list_tools 探测);
test 端点对单服务器隔离探测(连接失败降级为 ok=False,不抛)。
"""
from fastapi import APIRouter, HTTPException
from pydantic import TypeAdapter

from tools.mcp.client import MCPToolProvider
from tools.mcp.config import (
    MCPConfig,
    MCPServerConfig,
    add_server,
    load_mcp_config,
    remove_server,
)

router = APIRouter(prefix="/mcp", tags=["mcp"])

# Annotated 联合别名无 .model_validate,须用 TypeAdapter 解析 discriminator
_CONFIG_ADAPTER = TypeAdapter(MCPServerConfig)


@router.get("/servers")
def list_servers():
    cfg = load_mcp_config()
    return {"servers": {name: s.model_dump(exclude_none=True) for name, s in cfg.servers.items()}}


@router.post("/servers")
def create_server(name: str, config: dict):
    """新增服务器:name 为唯一键,config 为 StdioServer/HttpServer 配置(discriminator 解析)。"""
    try:
        server = _CONFIG_ADAPTER.validate_python(config)
    except Exception as e:
        raise HTTPException(status_code=422, detail=f"配置无效:{e}") from e
    add_server(name, server)
    return {"ok": True, "name": name}


@router.delete("/servers/{name}")
def delete_server(name: str):
    removed = remove_server(name)
    if not removed:
        raise HTTPException(status_code=404, detail=f"服务器 {name} 不存在")
    return {"ok": True, "name": name}


@router.post("/servers/{name}/test")
def test_server(name: str):
    """单服务器连通性探测:能发现工具即通;失败返回 ok=False(复用降级契约)。"""
    cfg = load_mcp_config()
    server = cfg.servers.get(name)
    if server is None:
        raise HTTPException(status_code=404, detail=f"服务器 {name} 不存在")
    single = MCPConfig(servers={name: server})
    found = MCPToolProvider(config=single, timeout=10).list_tools()
    return {"ok": bool(found), "tools": found, "name": name}
